osquery/events/windows/windowseventlogpublisher.cpp

```diff
@@ -0,0 +1,205 @@
+  }
+
+  const auto& character_frequencies_array =
+      config["feature_vectors"]["character_frequencies"];
+  if (character_frequencies_array.Empty()) {
+    return Status::failure(
+        "The character_frequencies configuration entity array is empty");
+  }
+
+  std::vector<double> output(kCharFreqVectorLen, 0.0);
+  for (unsigned int i = 0; i < character_frequencies_array.Size(); i++) {
+    if (character_frequencies_array[i].IsDouble()) {
+      output[i] = character_frequencies_array[i].GetDouble();
+    } else {
+      return Status::failure(
+          "The character_frequencies configuration entity array is not valid. "
+          "Entry #" +
+          std::to_string(i) + " is not a double");
+    }
+  }
+
+  character_frequency_map = std::move(output);
+  return Status::success();
+}
+} // namespace
+
+FLAG(bool,
+     enable_windows_events_publisher,
+     false,
+     "Enables the Windows events publisher");
+
+REGISTER(WindowsEventLogPublisher,
+         "event_publisher",
+         "WindowsEventLogPublisher");
+
+struct WindowsEventLogPublisher::PrivateData final {
+  std::vector<EvtSubscription::Ref> subscription_list;
+  std::shared_ptr<WindowsEventLogParserService> parser_service;
+};
+
+WindowsEventLogPublisher::WindowsEventLogPublisher() : d(new PrivateData) {}
+
+WindowsEventLogPublisher::~WindowsEventLogPublisher() {}
+
+void WindowsEventLogPublisher::configure() {
+  if (!FLAGS_enable_windows_events_publisher) {
+    return;
+  }
+
+  tearDown();
+
+  std::vector<double> character_frequency_map;
+  auto status = loadCharacterFrequencyMap(character_frequency_map);
+  if (!status.ok()) {
+    LOG(ERROR) << status.getMessage();
+  }
+
+  for (auto& current_subscriber : subscriptions_) {
+    auto sc = getSubscriptionContext(current_subscriber->context);
+    sc->character_frequency_map = character_frequency_map;
+
+    const auto& channel_list = sc->channel_list;
+    for (const auto& channel : channel_list) {
+      EvtSubscription::Ref subscription = {};
+
+      status = EvtSubscription::create(subscription, channel);
+      if (!status.ok()) {
+        auto error = GetLastError();
+        LOG(WARNING) << "Failed to subscribe to " << channel << ": " << error;
+
+      } else {
+        d->subscription_list.push_back(std::move(subscription));
+      }
+    }
+  }
+}
+
+Status WindowsEventLogPublisher::run() {
+  if (!FLAGS_enable_windows_events_publisher) {
+    return Status::failure("Publisher disabled by configuration");
+  }
+
+  d->parser_service = std::make_shared<WindowsEventLogParserService>();
+  Dispatcher::addService(d->parser_service);
+
+  while (!isEnding()) {
+    EvtSubscription::EventList event_list;
+
+    for (auto& subscription : d->subscription_list) {
+      auto event_list = subscription->getEvents();
+
+      if (!event_list.empty()) {
+        d->parser_service->addEventList(subscription->channel(), event_list);
+      }
+    }
+
+    std::this_thread::sleep_for(std::chrono::seconds(2U));
```

Comment:
This is too long of a wait. If osquery is shutting down you may exhaust the timeout and force the shutdown logic to force-exit instead of a clean exit. Is there an alternative to sleeping the thread? Can you poll/select/etc?

Reply:
The goal here was to collect an N-seconds worth of events to take advantage of write batching. I have updated this to remove the sleep and check with a timestamp instead.